### ISSUE TITLE
Problem in Bootstrapping

### ISSUE THREAD

**Issue by fawad83** [sem-in-r/seminr#154]
Hi

I am trying to run the following code but i am encountering the error 

#SEMinR
library(seminr)
library("readxl")
DSForR <- read_excel("DSForR.xlsx")
inn_mm <- constructs(
  reflective("INN_P", multi_items("INN_P", 1:4)),
  reflective("INN_M", multi_items("INN_M", 1:4))
)
structure <- relationships(paths(from = "INN_P", to = "INN_M"))
pls <- estimate_pls(data = DSForR, inn_mm, structure)
summary(pls)
model_summary <- summary(pls)
model_summary$iterations
model_summary$cross_loadings
model_summary$loadings
model_summary$t_values
model_summary$p_values

boot_estimates <- bootstrap_model(seminr_model = pls,
                                 nboot = 1000,
                                 cores = 2)
summary(boot_estimates)
boot_model_summary$nboot

Bootstrapping model using seminr...
Bootstrapping encountered this ERROR: 
'x' must be an array of at least two dimensions> summary(boot_estimates)
Length  Class   Mode 
     0   NULL   NULL 
> boot_model_summary$nboot
Error: object 'boot_model_summary' not found

**Comment by soumyaray** [comment]
Hi Fawad, just want to confirm that we are able to replicate the error using your code but our own data. Will get back to you as we make progress.

unrelated note: try to avoid the variable name  `structure`, as that is the name of a function in R already. It likely won't affect your code but could cause a surprise later.

**Comment by fawad83** [comment]
Thanks so much. I appreciate your time and support.

**Comment by soumyaray** [comment]
Fawad, it seems our bootstrapping didn't work well for very simple models with only a single antecedent – I don't know how no one caught this before!!

I believe we have a working fix now. Your options are:
- Just avoid models with only one antecedent (I'm guessing it was a toy model)
- Wait for an upcoming patch (though we might wait to fix some other little things that folks report)
- Or else we will can give you instructions on how to side-load the fixed code from Github before an official release

I'll let @NicholasDanks to review my fix and decide when to release a patch

**Comment by fawad83** [comment]
Dear Prof. Soumya
Thanks for letting me know. I have recently started learning R. Yes i was just a try. Let me work on it and make a more complex model. 

I will be waiting for update on the SEMinR ver 1.1.0.

Thanks for your time. Really Appreciate it.

**Comment by fawad83** [comment]
Hi

I just tried to run the model with multiple IVs but now i am getting a different error. 

Bootstrapping model using seminr...
Bootstrapping encountered this ERROR: 
incorrect number of dimensions> summary(boot_estimates)

This is my code

#SEMinR
library(seminr)
library("readxl")
DSForR <- read_excel("DSForR.xlsx")
inn_mm <- constructs(
  reflective("INN_P", multi_items("INN_P", 1:4)),
  reflective("INN_M1", multi_items("INN_M", 1:4)),
  reflective("INN_M2", multi_items("INN_M", 5:8))
)
stm <- relationships(paths(from = "INN_P", to = c("INN_M1","INN_M2")))
pls <- estimate_pls(data = DSForR, inn_mm, stm)
summary(pls)
model_summary <- summary(pls)
model_summary$iterations
model_summary$cross_loadings
model_summary$loadings
model_summary$t_values
model_summary$p_values

boot_estimates <- bootstrap_model(pls, nboot = 1000, cores = 2)
summary(boot_estimates)

**Comment by soumyaray** [comment]
Yes, I believe we caught this as well (again, one antecedent only) and fixed it as well. If you are only playing around for now, I suggest a model with two antecedents to one (or more) outcomes:

```
stm <- relationships(paths(from = c("INN_P", "INN_M1"), to = "INN_M2"))
```

I'm guessing the two INN_M* constructs are highly correlated, so it may not produce nice results :D

**Comment by soumyaray** [comment]
I'll double check later, though

**Comment by fawad83** [comment]
Wow. It ran. Thanks, sorry silly mistake. From should have been two
antecedent.

Thanks Prof. Appreciate your patience and time. Will now read in more
detail and will ask for help if needed.

Stay Blessed and Stay Safe.

**Comment by soumyaray** [comment]
Best of luck!  Don't hesitate to ask questiosn either on our FB group (general PLS questions) or here (possible bugs or feature requests).

**Comment by fawad83** [comment]
Will Surely do. Thanks.

**Comment by NicholasDanks** [comment]
Corrected by PR #157

**Comment by fawad83** [comment]
Hi Prof.

I hope you are keeping well. I am learning SEMinR. I normally use
SMART-PLS. Now while using SEMinR, i noticed that loadings, AVE values are
different between SMART-PLS and SEMinR. I understand the bootstrapping
results also might differ since they are based on random numbers.

**Comment by soumyaray** [comment]
Hi @fawad83,  please see my response to #131 to check if its the same issue. Briefly, try using `composite()` instead of `reflective()` to get the same results as SmartPLS. If my comments to #131 do not address your concern, please start a post a new issue!  And please supply some code for us to try and replicate results. Many thanks for considering SEMinR and hope to hear from you.

**Comment by fawad83** [comment]
Dear Prof.

Thanks for your response. I will surely check. One question though,

Composite is for formative constructs, but what if our constructs are
reflective. So if we use Mode A with composite, it will be treated as
reflective?

**Comment by soumyaray** [comment]
@fawad83 SmartPLS uses the terms 'reflective' and 'formative' in different ways than the methodological literature, which is the source of the confusion.  PLS models generally cannot estimate either reflective or formative types of constructs. Both these types of constructs in SmartPLS (or any regular PLS software) are just composites (mode A and mode B, respectively, which denote different types of weighting). SmartPLS believes that mode A and mode B approximate reflective and formative constructs, but this belief is not generally shared by others.

If you wish to do true "reflective" modeling, then you have two options:

a) Use consistent-PLS (PLSc) which adjusts the PLS estimates to better approximate the reflective constructs. You can use our `reflective()` function, which will apply consistent-PLS (PLSc) to the results. SmartPLS also has a consistent-PLS mode that you have to specially apply.

b) Use a covariance-based SEM (CBSEM) modeling techniques that are purpose-made for reflective constructs. You can estimate your model using CBSEM in SEMinR using the `estimate_cbsem()` function after defining all your constructs as `reflective()`. Or, you can use dedicated tools for CBSEM such as LISREL/AMOS.

Hope that helps.

**Comment by fawad83** [comment]
Thanks Prof. A lot clear now. Will ask if i have any confusion.

**Comment by soumyaray** [comment]
Sure thing @fawad83 -- I encourage you to either post to a new issue, or to a closed by relevant issue (ilke #131) instead of here. We hope this discussion can help others but they likely won't look under an issue about bootstrapping about it.

**Comment by fawad83** [comment]
Sure Prof. Thanks again for your time.
